### DISCUSSION TITLE
Issue with http request to external, e.g. market place / plugins

### DISCUSSION THREAD

**Discussion by jasperdejong-trell** [kimai/kimai#4661] [On-Premise: Installation & Updates] [answered]
Hi, 
I have a working on prem installation on NGINX + php8.3-fpm. 
However, when navigation to "System -> Plugins", it gives me a 500 error. 
Going through the logs, there is a timeout thrown when connecting to the external. it looks like: 
`[request] CRITICAL: Uncaught PHP Exception Symfony\Component\HttpClient\Exception\TimeoutException: "Idle timeout reached for "https://www.kimai.org/plugins.json"." at ErrorChunk.php line 56`

I have a strong feeling this has to do with either our proxy server, or a SSL error, due to the proxy. 
Is there a way to set a proxy server for the application? 
Or get more debug information?

Thanks

**Comment by kevinpapst** [comment]
Sorry for that stupid mistake, I thought errors were catched, but seems not all of them. 

You could patch according to this commit to check if the 500 goes away:
https://github.com/kimai/kimai/pull/4659/commits/ccaa341bd97173715267c049f301a1c147ece2bb

Some docs to read: https://symfony.com/doc/current/http_client.html#http-proxies
So setting `https_proxy` should be enough I hope.

You could try to add a global `proxy` URL as configuration in config/packages/local.yaml (create it, if it does not exist):
https://symfony.com/doc/current/http_client.html#configuration

**Reply by jasperdejong-trell** [reply]
thanks, first of all, no need to say sorry ;-)
However, the proxy setting does not change, hence I believe it's now a matter of SSL/TLS. For the test I would like to disable verify_host and verify_peer in the http_client. But (sorry for my lacking of knowledge of the framework) where do I need to set it?

**Reply by kevinpapst** [reply]
As I wrote 

> You could try to add a global proxy URL as configuration in config/packages/local.yaml (create it, if it does not exist):

Create `config/packages/local.yaml` - see https://www.kimai.org/documentation/local-yaml.html

I don't know either, should be something like the code below, you will see an error [upon cache reload](https://www.kimai.org/documentation/cache.html) if it is wrong.

Untested:
```yaml
# config/packages/local.yaml
framework:
    http_client:
        default_options:
            verify_host: true
            verify_peer: true
```

Edit: Please let me know what you did, so we can add it to the docs

**Reply by jasperdejong-trell** [reply]
Hi Kevin, 
I'll for sure update here once I have the answer. 
I've narrowed it down to Openssl, we use Zscaler proxies with TLS/SSL inspection, hence this has consequences to the certificates....
So I have to dive into it a bit more, so could take another day before I'm able to fix the issue 
(ps. it's running on a DEBIAN server)

**Reply by jasperdejong-trell** [reply]
Hi, 
A small update, however sadly enough no solution yet. 
It seems that its openssl related, at least it boils down to 'certificates'. 
The 'veryify_peer' option does not change a thing. 
I've been testing with command line openssl (client). A normal openssl client session returns an error that 'local issuer cannot be found' when connecting to e.g. 'kimai.org'. But, when I explicit tell openssl which CA certificate to use (zscaler) with the -CAfile option and point it to the zScalerRootCA, the connection is ok. 
Meaning that this is most probably the source of the issue. However, the problem is now, how to get php and/or kimai to use these settings. For the problematic 'get_file_contents' function, I would need to set the SSL options in the context. Like: 
```
$stream_context = stream_context_create([ 'ssl' => [
'local_cert'        => '/path/to/key.pem',
'peer_fingerprint'  => openssl_x509_fingerprint(file_get_contents('/path/to/key.crt')),
'verify_peer'       => false,
'verify_peer_name'  => false,
'allow_self_signed' => true,
'verify_depth'      => 0 ]]);
```
But I'm not sure if this is the same context which the httpClient in symfony uses?

**Reply by kevinpapst** [reply]
Never tried, but I read the docs that it is.

The 
https://symfony.com/doc/current/reference/configuration/framework.html#http-client 
allows to set e.g. the 
https://symfony.com/doc/current/reference/configuration/framework.html#local-cert
https://symfony.com/doc/current/reference/configuration/framework.html#peer-fingerprint

So I assume these options are passed down 1:1.

**Reply by jasperdejong-trell** [reply]
I'll go through them. 
Which file in the source is there to alter in this case?

**Reply by jasperdejong-trell** [reply]
Thanks for being this available Kevin! highly appreciated!

**Reply by kevinpapst** [reply]
> Which file in the source is there to alter in this case?

I already answered that above: `config/packages/local.yaml` is the best place to add these settings

**Reply by jasperdejong-trell** [reply]
I'll start testing with the ssl part of the httpclient and come back with the outcome

**Reply by jasperdejong-trell** [reply]
IT WORKS!! 
![image](https://github.com/kimai/kimai/assets/154524871/34179568-369b-48bf-b0df-35301d3b9631)

**Comment by jasperdejong-trell** [comment]
It's not 100% correct yet, since I have a local proxy setup now and used proxy-chaining, since I was hoping the local proxy would honor the certificates. However, I'll revert my configuration of the server back to the normal "production" configuration and test again. Once I have it working, I'll update you with the information for the docs ;-)

It should not make a huge difference though, since it's last proxy in the chain which is causing this behavior. Anyway, tomorrow I'll give you the correct config for this scenario..

**Reply by kevinpapst** [reply]
You love fiddling around, don't you 😁 

Are you sure the additional config settings don't work? If you got it working with your previous PHP snippet, then it should work with the Symfony configuration in local.yaml as well.

**Reply by jasperdejong-trell** [reply]
I do! And can't stand when I don't have the knowledge of something...hence I dive into it :-)
See below the complete routine for the setup with SSL inspecting proxies, please let me know in case of any questions, or missing parts....happy to support!

**Reply by kevinpapst** [reply]
> I do! And can't stand when I don't have the knowledge of something...hence I dive into it :-)

I know that feeling 😆

**Comment by jasperdejong-trell** [comment, marked as answer]
**Kimai configuration/ setup with proxy+SSL inspection**

In case your network configuration is setup in a proxy environment with SSL inspection, you might run into issue where some functionality may not work by default. 
It’s possible that you encounter an “Error 500” when loading modules with an external interface (e.g. “System -> Plugins” which loads external content). When examining the logfiles you would see a message like “[request] CRITICAL: Uncaught PHP Exception Symfony\Component\HttpClient\Exception\TimeoutException: "Idle timeout reached for "https://www.kimai.org/plugins.json"." at ErrorChunk.php line 56” 
or a reference to “get_file_contents” with a similar timeout message. 
In order to fix the issue, there a several steps to take: 
For Linux (Debian base) users:
1.	export the RootCA certificate from your inspecting proxy (e.g. zScalerRootCAcrt), there are many guidelines available which explain how to export from a browser.
1.	Copy your CA to dir /usr/local/share/ca-certificates/
2.	Use command: sudo cp foo.crt /usr/local/share/ca-certificates/foo.crt
3.	Update the CA store: sudo update-ca-certificates
2.	Make sure the ‘etc/ssl/certs’ dir has the correct permissions:
a.	Sudo chmod -R 775 /etc/ssl/certs
3.	Next step is to create a ‘local.yaml’ file in the <path to your www document root>/kimai/config/packages directory (e.g. sudo nano /var/www/html/kimai/config/packages/local.yaml)
4.	Add the proxy and certificate to the HttpClient configuration:

# config/packages/local.yaml
```yaml
framework:
    http_client:
        default_options:
            proxy: http://127.0.0.1:3128
            cafile: "/etc/ssl/certs/zScalerRootCA.crt"
#            verify_host: false
#            verify_peer: false
```

5.	The last two lines can be used to test before you set the cafile option. 
6.     (sudo) bin/composer kimai:reload

- sudo chown -R www-data:www-data . (or your webserver user)

- sudo chmod -R g+r .

- sudo chmod -R g+rw var/

Done, you can now use Kimai with external interfacing....

**Reply by kevinpapst** [reply]
Well done, thanks for sharing ❤️ 

The local.yaml won't be touched during updates, so its safe for future use.
This is worth sharing in the documentation, I just need to find a place where it fits - will leave a link once I did.

**Reply by jasperdejong-trell** [reply]
Thank you for the directions! 
No thanks needed, it's good to give something back to the project. 

Again, anything you miss / need to have changed, just let me know!
